### PR TITLE
redo eventHub client facade to enable mocking

### PR DESCRIPTION
**Pull request by SreeramGarlapati**
addresses following issues filed: #209 #169 #183 #222 #221

### CODE REVIEW

**Comment by SreeramGarlapati** [comment]
> public class AmqpException extends Exception {

see if we can get this out

---
Refers to: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/AmqpException.java:12 in 67149cd. [](commit_id = 67149cd324a0d86fc76df7eb6dfd009cc5d11381, deletion_comment = False)

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/EventHubPartitionRuntimeInformation.java [hunk 1]
```diff
@@ -6,57 +6,17 @@
...
-    public String getLastEnqueuedOffset() {
-        return this.lastEnqueuedOffset;
-    }
-
-    public Instant getLastEnqueuedTimeUtc() {
-        return this.lastEnqueuedTimeUtc;
-    }
+public interface EventHubPartitionRuntimeInformation {
+
+    String getEventHubPath();
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/EventHubPartitionRuntimeInformation.java]: >getEventHubPath [](start = 11, length = 15)

do we want to call this name? breaking change.!

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/EventHubRuntimeInformation.java [hunk 2]
```diff
@@ -6,39 +6,13 @@
...
-    final String[] partitionIds;
+public interface EventHubRuntimeInformation {
 
-    EventHubRuntimeInformation(final String path, final Instant createdAt, final int partitionCount, final String[] partitionIds) {
-        this.path = path;
-        this.createdAt = createdAt;
-        this.partitionCount = partitionCount;
-        this.partitionIds = partitionIds;
-    }
+    String getPath();
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/EventHubRuntimeInformation.java]: >getPath [](start = 11, length = 7)

getName? breaking change?

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/ReceiverRuntimeInformation.java [hunk 3]
```diff
@@ -10,76 +10,40 @@
...
     /**
      * Get sequence number of the {@link EventData}, that is written at the end of the Partition Stream.
      *
      * @return last sequence number
      */
-    public long getLastSequenceNumber() {
-
-        return this.lastSequenceNumber;
-    }
+    long getLastSequenceNumber();
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/ReceiverRuntimeInformation.java]: >getLastSequenceNumber [](start = 9, length = 21)

copy-pasted from .net
Should this have been getLast**Enqueued**SequenceNumber ???? copy-pasted from .net

**Comment by SreeramGarlapati** [comment]
@JonathanGiles / @osgigeek / @ialexandrov / @adoelger / @bojingo / @wofr / @yoshioterada and all other consumers/advisors of the API:

We made some changes to the API for the upcoming 1.0 release - to address these 2 goals:
1. improve discoverability of our API surface
2. enable mock'ability of our APIs. 

Here's what we did:

1. moved all code that's not needed by the consumer of our library from our main package to `impl` package
2. exposed just the bare-minimum - explicit contract of our APIs in the main package (the `com.microsoft.azure.eventhubs`)

Here's how the eventhubs java client API looks like with this refactoring:

```
// EventHubClient used to be a class - now its an interface with a factory method to create the client
final EventHubClient eventhubClient = EventHubClient.createFromConnectionStringSync(
                        "----ConnectionStringFromAzurePortal----",
                        executor);
```
```
// Changed from new EventData() to EventData.create()
// EventData is an interface and has a factory method which returns our specific implementation - EventDataImpl
eventhubClient.sendSync(EventData.create("deserialized_data_bytes")); 
```
```
// PartitionReceiver changed from class to a mockable interface
final PartitionReceiver receiver = eventhubClient.createReceiverSync(
        	EventHubClient.DEFAULT_CONSUMER_GROUP_NAME,
                "0", 
                EventPosition.fromStartOfStream());
```
```
// return type or receive API changed from Iterable<EventData> to Iterable<? extends EventData>
final Iterable<? extends EventData> events = receiver.receiveSync(MAX_EVENTS_PER_BATCH);
```

Let us know how it looks...

**Comment by SreeramGarlapati** [comment]
> public class AmqpException extends Exception {

will review and address the exception contract in an upcoming PR. I want to confine the goal of this PR to "enabling mockability".

---
In reply to: [363868947](https://github.com/Azure/azure-event-hubs-java/pull/244#issuecomment-363868947) [](ancestors = 363868947)

---
Refers to: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/amqp/AmqpException.java:12 in 67149cd. [](commit_id = 67149cd324a0d86fc76df7eb6dfd009cc5d11381, deletion_comment = False)

**Comment by nathan-gs** [comment]
> EventDataImpl; does this sound awkward ? 
Yes it does. Simple fix is to get rid of `EventDataImpl` and use `EventData` everywhere.

**Comment by nathan-gs** [comment]
Great initiative!

**Comment by yarinm** [comment]
This is definitely looks like the right direction, using interfaces is the way to go. *thumbsup*

**Pull request merged**